File: score_lda.R
suppressPackageStartupMessages({
  library(MASS)
  library(readr)
  library(stringr)
  library(dplyr)
  library(purrr)
  library(lubridate)
  library(argparse)
})

timestamp_started <- now()

parser <- ArgumentParser()
parser$add_argument("-s", "--seed", type="integer", default=451,
    help="Random seed")

parser$add_argument("-p", "--pairs", type="integer", default=200,
    help="Number of gene pairs to pick as relevant from a component")

parser$add_argument("-b", "--bootstrap", type="double", default=0.66,
    help="Fraction of genes to bootstrap")

parser$add_argument("-n", "--name", type="character",
    help="Name of cell type to target")

parser$add_argument("-t", "--type", type="character",
    help="General or specific cell type")


parser$add_argument("-i", "--input", type="character",
    default=file.path("results", "gene_data_vs_cell_type.tsv"),
    help="Path to input data")

parser$add_argument("-o", "--output", type="character",
    default=file.path("results", "results_pairs.tsv"),
    help="Path to output pair/scores data")

parser$add_argument("-f", "--feedback", type="character",
    default=file.path("results", "results_feedback.tsv"),
    help="Path to output feedback on run. Errors, time, etc")

args <- parser$parse_args()

random_seed <- args$seed
# random_seed <- 451
fraction_bootstrap <- args$bootstrap
# fraction_bootstrap <- 0.66
top_genes <- ceiling(1/2 * (sqrt(8*args$pairs + 1) + 1)) # this uses nC2 to pick the top set to create ~number of pairs
# top_genes <- ceiling(1/2 * (sqrt(8*200 + 1) + 1))
input_path <- args$input
# input_path <- file.path("results", "gene_data_vs_cell_type.tsv")
cell_name <- args$name %>% str_replace_all('"','')
# cell_name <- '"Monocyte"' %>% str_replace_all('"','')
cell_type <- args$type
# cell_type <- "General_Cell_Type"
input_path <- args$input
# input_path <- file.path("results", "gene_data_vs_cell_type.tsv")
output_path <- args$output
# output_path <- file.path("results", "results.tsv")
feedback_path <- args$feedback
# feedback_path <- file.path("results", "feedback.tsv")



# function to note statistics on done
# this will stop the script
note_complete <- function(success, message) {
  data.frame(
      start = timestamp_started,
      end = now(),
      cell_name = cell_name,
      cell_type = cell_type,
      random_seed = random_seed,
      fraction_bootstrap = fraction_bootstrap,
      top_genes = top_genes,
      input_path = input_path,
      output_path = output_path,
      success = success,
      message = message
    ) %>%
    write_tsv(feedback_path)
  quit()
}

# Read base data file and transform for type
input_path %>%
  read_tsv(col_types=cols(
    .default = col_double(),
    GSM_ID = col_character(),
    Cell_Type = col_character(),
    General_Cell_Type = col_character()
  )) %>%
  mutate_(.dots = setNames(cell_type, "type_truth")) %>%
  select(-GSM_ID, -Cell_Type, -General_Cell_Type) ->
  genes

genes %>% select(-type_truth) -> gene_data
genes %>%
  select(type_truth) %>%
  unlist %>%
  unname %>%
  map_chr(~ if (. != cell_name) { "other" } else { cell_name }) ->
  gene_labels

set.seed(random_seed)

# Bootstrap genes, generate gene indexes
gene_data %>%
  ncol %>%
  `*`(fraction_bootstrap) %>%
  floor ->
  genes_count
gene_data %>%
  ncol %>%
  sample(replace = TRUE) %>%
  head(genes_count) %>%
  select(gene_data, .) %>%
  cbind(data.frame(type_truth = gene_labels)) ->
  genes_bootstrapped

# Perform downsampling to match given set
genes_bootstrapped %>%
  filter(type_truth == cell_name) ->
  genes_bootstrapped_truth
genes_bootstrapped %>%
  filter(gene_labels != cell_name) %>%
  sample_n(nrow(genes_bootstrapped_truth)) %>%
  rbind(genes_bootstrapped_truth) ->
  genes_bootstrapped_downsampled

# LDA fit for components
lda_fit <- function(df) {
  tryCatch({
    results <- lda(df %>%
                     select(-type_truth) %>%
                     as.matrix,
                   df %>%
                     select(type_truth) %>%
                     unlist %>%
                     unname %>%
                     as.character,
                   CV = FALSE)
    results
  }, error = function(errorCondition) {
    # Note this will force the script to stop
    note_complete(FALSE, errorCondition$message)
  })
}

# Select top loading genes from components
genes_bootstrapped_downsampled %>%
  lda_fit %>%
  get("scaling", .) %>%
  (function(m) {
    data.frame(
      loading = m %>% unlist %>% unname,
      gene = m %@% "dimnames" %>% nth(1)
    )
  }) %>%
  mutate(
    loading = abs(loading),
    gene = as.character(gene)
  ) %>%
  arrange(-loading) %>%
  head(top_genes) %>%
  get("gene", .) %>%
  unlist ->
  top_gene_names

# Perform Leave One Out
data_range <- 1:nrow(genes_bootstrapped_downsampled)
data_range %>%
  map(function(idx){
    # Create data sets
    data_train <- genes_bootstrapped_downsampled %>% filter(idx != data_range)
    data_test <- genes_bootstrapped_downsampled %>% filter(idx == data_range)

    lda_model <- lda_fit(data_train)

    data_test %>%
      select(type_truth) %>%
      unlist %>%
      unname %>%
      as.character ->
      type_truth

    data_test %>%
      select(-type_truth) %>%
      predict(lda_model, newdata = .) %>%
      get("class", .) ->
      type_predicted

    data.frame(
      idx = idx,
      type_truth = type_truth,
      type_predicted = type_predicted
    )
  }) %>%
  reduce(rbind) %>%
  as.data.frame ->
  data_results

# Score is simply accuracy of the LOO predictions between the balanced classes
#  of the type and other
data_results %>%
  mutate(correct = type_truth == type_predicted) %>%
  select(correct) %>%
  unlist %>%
  sum %>%
  `/`(nrow(data_results)) ->
  score

# Pairs a vector into low/high pairs, with a frequency of one
pair_ <- function(v) {
  if (length(v) < 2) {
    return(data.frame())
  }
  v %>% head(1) -> h
  v %>% tail(-1) -> t
  h %>%
    rep(length(t)) %>%
    data.frame(low=., high=t, frequency=as.integer(1)) %>%
    rbind(pair_(t))
}
pair <- function(v) {
  v %>%
    as.character %>%
    sort %>%
    pair_
}

top_gene_names %>%
  pair %>%
  mutate(
    score = score,
    cell_name = cell_name,
    cell_type = cell_type
  ) %>%
  write_tsv(output_path)

note_complete(TRUE, NA)

File: score_lda.py
import argparse
import math
from itertools import combinations
import numpy as np
import pandas as pd
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.preprocessing import StandardScaler


parser = argparse.ArgumentParser(description='Perform LDA')
parser.add_argument('--seed', type=int, default=451,
                    help='Random seed for operation')
parser.add_argument('--pairs', type=int, default=200,
                    help='Number of gene pairs to pick as relevant from a component')
parser.add_argument('--bootstrap', type=float, default=0.66,
                    help='Fraction of genes to bootstrap')
parser.add_argument('--name', type=str, required=True,
                    help='Name of cell type to target')
parser.add_argument('--type', type=str, required=True,
                    help='General or specific cell type')

parser.add_argument('--input', type=str, required=True,
                    help='Path to input data')
parser.add_argument('--output', type=str, required=True,
                    help='Path to output pair/scores data')

args = parser.parse_args()
# args = parser.parse_args(([
#     '--name', '450',
#     '--input', 'results/gene_data_vs_cell_type.tsv',
#     '--seed', '0',
#     '--type', 'General_Cell_Type',
#     '--name', '"Monocyte"',
#     '--output', 'results/score.00000.General_Cell_Type.Monocyte.tsv'
# ]))

# Important since some names have spaces
cell_name = args.name.replace('"', '')

data_00 = pd.read_table(args.input)
data_01 = data_00.assign(truth=data_00.Cell_Type if args.type ==
                         'Cell_Type' else data_00.General_Cell_Type) \
    .drop(['GSM_ID', 'Cell_Type', 'General_Cell_Type'], axis=1)

others = (data_01.truth == cell_name).apply(
    lambda x: cell_name if x else "other")
data_02 = data_01.assign(truth=others)

data_normal = data_02.query('truth != "other"').reset_index(drop=True)
data_other = data_02.query('truth == "other"')\
    .sample(n=data_normal.shape[0], random_state=args.seed)\
    .reset_index(drop=True)

data_03 = pd.DataFrame.append(data_normal, data_other)\
    .sample(frac=1, random_state=args.seed + 1)\
    .reset_index(drop=True)


def lda_model(data):
    X_train = data.drop(['truth'], axis=1).as_matrix()
    y_train = data[['truth']].as_matrix().T[0]

    scaler = StandardScaler()
    scaler.fit(X_train)
    X_train_scaled = scaler.transform(X_train)

    clf = LinearDiscriminantAnalysis()
    clf.fit(X_train_scaled, y_train)
    return clf, scaler


lda_model_main, _ = lda_model(data_03)
top_genes = math.floor(1 / 2 * (math.sqrt(8 * args.pairs + 1) + 1))

top_gene_names = pd.DataFrame({
    'gene': data_03.columns[:-1],
    'loading': np.abs(lda_model_main.scalings_[:, 0])})\
    .sort_values('loading', ascending=False)\
    .reset_index(drop=True)\
    .iloc[:top_genes]


def ignore_one(df, idx):
    skip_list = [i for i in range(0, df.shape[0]) if i != idx]
    return df.ix[skip_list]


def keep_one(df, idx):
    skip_list = [i for i in range(0, df.shape[0]) if i == idx]
    return df.ix[skip_list]


def leave_one_out(df, idx):
    data_train = ignore_one(df, idx)
    data_test = keep_one(df, idx)
    clf, scaler = lda_model(data_train)

    X_test = scaler.transform(
        data_test.drop(['truth'], axis=1))
    y_test = data_test[['truth']].as_matrix().T[0]
    result = clf.predict(X_test)
    return result[0] == y_test[0]


def run(data):
    return list(map(lambda idx: leave_one_out(data, idx),
                    range(0, data.shape[0])))

results = run(data_03)
score = np.sum(results) / len(results)


gene_pairs = np.array(list(combinations(np.sort(
    top_gene_names.gene, axis=-1, kind='quicksort', order=None), 2))).T

final_results = pd.DataFrame({
    'low': gene_pairs[0],
    'high': gene_pairs[1],
    'frequency': 1,
    'score': score,
    'cell_name': cell_name,
    'cell_type': args.type})[[
        'low',
        'high',
        'frequency',
        'score',
        'cell_name',
        'cell_type']]

final_results.to_csv(args.output, sep='\t',
                     encoding='utf-8', index=False)
